tests/sklearn/test_sklearn_models.py

```diff
@@ -474,75 +516,119 @@ def check_subfunctions(fitted_model, model_class, x):
     # not inherit from skorch.
     # FIXME: https://github.com/zama-ai/concrete-ml-internal/issues/3373
-    if get_model_name(fitted_model) == "NeuralNetRegressor":
+    if get_model_name(model) == "NeuralNetRegressor":
         with pytest.raises(
             NotImplementedError,
             match=(
                 "The `predict_proba` method is not implemented for neural network regressors. "
                 "Please call `predict` instead."
             ),
         ):
-            fitted_model.predict_proba(x)
+            model.predict_proba(x)
 
-    if get_model_name(fitted_model) == "KNeighborsClassifier":
-        # FIXME: https://github.com/zama-ai/concrete-ml-internal/issues/3962
-        pytest.skip("Skipping subfunctions test for KNN, doesn't work for now")
+    # KNeighborsClassifier does not provide a predict_proba method for now
+    # FIXME: https://github.com/zama-ai/concrete-ml-internal/issues/3962
+    elif get_model_name(model) == "KNeighborsClassifier":
+        with pytest.raises(
+            NotImplementedError,
+            match=(
+                "The `predict_proba` method is not implemented for KNeighborsClassifier. "
+                "Please call `predict` instead."
+            ),
+        ):
+            model.predict_proba(x)
 
-    if is_classifier_or_partial_classifier(model_class):
+    elif is_classifier_or_partial_classifier(model_class):
+
+        # Check outputs from the 'predict_proba' method (for all classifiers,
+        # except KNeighborsClassifier)
+        y_proba_clear = model.predict_proba(x)
+        y_proba_simulated = model.predict_proba(x, fhe="simulate")
 
-        fitted_model.predict_proba(x)
+        assert y_proba_clear.shape == y_proba_simulated.shape, (
+            "Method 'predict_proba' from Concrete ML outputs different shapes when executed"
+            "in the clear and with simulation."
+        )
+        check_float_array_equal(y_proba_clear, y_proba_simulated)
 
-        # Only linear classifiers have a decision function method
         if is_model_class_in_a_list(model_class, _get_sklearn_linear_models()):
-            fitted_model.decision_function(x)
+
+            # Check outputs from the 'decision_function' method (for all linear classifiers)
+            y_scores_clear = model.decision_function(x)
+            y_scores_simulated = model.decision_function(x, fhe="simulate")
+
+            assert y_scores_clear.shape == y_scores_simulated.shape, (
+                "Method 'decision_function' from Concrete ML outputs different shapes when executed"
+                "in the clear and with simulation."
+            )
+            check_float_array_equal(y_scores_clear, y_scores_simulated)
 
 
-def check_subfunctions_in_fhe(model, fhe_circuit, x):
-    """Check subfunctions in FHE: calls and correctness."""
+def check_separated_inference(model, fhe_circuit, x, check_float_array_equal):
```

Comment:
this test is this Pr's initial goal : it tests that all inference methods (predict, predict_proba, decision_function) match the expected results using the circuit's ouputs + other additional computations (post_processing, argmax)